R/print.data.table.R

```diff
@@ -164,3 +189,39 @@ shouldPrint = function(x) {
 #   as opposed to printing a blank line, for excluding col.names per PR #1483
 cut_top = function(x) cat(capture.output(x)[-1L], sep = '\n')
 
+# to calculate widths of data.table for PR #4074
+# gets the width of the data.table at each column
+#   and compares it to the console width
+dt_width = function(x, class, row.names, col.names) {
+  widths = apply(nchar(x, type='width'), 2L, max)
+  if (class) widths = pmax(widths, 6L)
+  if (col.names != "none") names = sapply(colnames(x), nchar, type = "width") else names = 0L
+  dt_widths = pmax(widths, names)
+  rownum_width = if (row.names) as.integer(ceiling(log10(nrow(x)))+1) else 0L
+  cumsum(dt_widths + 1L) + rownum_width
+}
+# keeps the dim and dimnames attributes
+toprint_subset = function(x, cols_to_print) {
+  if (nrow(x) == 1L){
+    atts = attributes(x)
+    atts$dim = c(1L, sum(cols_to_print))
+    atts$dimnames[[2L]] = atts$dimnames[[2L]][cols_to_print]
+    x = x[, cols_to_print, drop=FALSE]
+    attributes(x) = atts
+    x
+  } else {
+    x[, cols_to_print, drop=FALSE]
+  }
+}
+# message for when trunc.cols=TRUE and some columns are not printed
+trunc_cols_message = function(not_printed, abbs, class){
+  n = length(not_printed)
+  if (class) classes = paste0(" ", tail(abbs, n)) else classes = ""
+  not_printed_paste = paste0(not_printed, classes, collapse = ", ")
+  cat(sprintf(ngettext(n,
+                       paste0("1 variable not shown: %s"),
+                       paste0(n, " variables not shown: %s")),
```

Comment:
Last thing: for #3935, `ngettext` will work best if we just pass plain formatting strings:

```
ngettext(n, "%d variable not shown: %s",
                       "%d variables not shown: %s"),
              n, not_printed_paste)
```

Reply:
Yes, I like that. And if it makes it easier for the translation, I'm all for it. It was the reason I used `ngettext()` in the first place but had rarely used it before.